File: design-patterns/class-03.09/payments/credit-card-factory.go
package payments

import (
	"errors"
	"regexp"
)

type CreditCardFactory struct {
	cardNumber string
}

func NewCreditCardFactory(cardNumber string) *CreditCardFactory {
	return &CreditCardFactory{cardNumber: cardNumber}
}

func (f *CreditCardFactory) CreatePayment() (Payment, error) {
	match, _ := regexp.MatchString(`^\d{16}$`, f.cardNumber)

	if !match {
		return nil, errors.New("número de cartão inválido: deve conter exatamente 16 dígitos")
	}

	return CreditCardPayment{cardNumber: f.cardNumber}, nil
}

File: design-patterns/class-25.08/01.exercise-payments/CreditCardFactory.java
public class CreditCardFactory extends PaymentFactory {
  private String cardNumber;

  public CreditCardFactory(String cardNumber) {
    this.cardNumber = cardNumber;
  }

  @Override
  public Payment createPayment() {
    if (cardNumber == null || !cardNumber.matches("\\d{16}")) {
      throw new IllegalArgumentException("Número de cartão inválido: deve conter exatamente 16 dígitos.");
    }

    return new CreditCardPayment(this.cardNumber);
  }
}
